vsphere_plugin_common/__init__.py

```diff
@@ -1662,6 +1674,57 @@ def stop_server(self, server):
         self._wait_for_task(task)
         logger().debug("Server is now stopped.")
 
+    def backup_server(self, server, snapshot_name):
+        if server.obj.snapshot:
+            snapshot = self.get_snapshot_by_name(
+                server.obj.snapshot.rootSnapshotList, snapshot_name)
+            if snapshot:
+                raise NonRecoverableError(
+                    "Snapshot {snapshot_name} already exists."
+                    .format(snapshot_name=snapshot_name,))
+
+        task = server.obj.CreateSnapshot(
+            snapshot_name, description="Backup created by vsphere plugin.",
+            memory=False, quiesce=False)
+        self._wait_for_task(task)
+
+    def get_snapshot_by_name(self, snapshots, snapshot_name):
+        for snapshot in snapshots:
+            if snapshot.name == snapshot_name:
+                return snapshot
+            else:
+                return self.get_snapshot_by_name(snapshot.childSnapshotList,
+                                                 snapshot_name)
+        return False
+
+    def restore_server(self, server, snapshot_name):
+        if server.obj.snapshot:
+            snapshot = self.get_snapshot_by_name(
+                server.obj.snapshot.rootSnapshotList, snapshot_name)
+        else:
+            snapshot = None
+        if not snapshot:
+            raise NonRecoverableError(
+                "No snapshots found with name: {snapshot_name}."
+                .format(snapshot_name=snapshot_name,))
+
+        task = snapshot.snapshot.RevertToSnapshot_Task()
+        self._wait_for_task(task)
+
+    def remove_backup(self, server, snapshot_name):
+        if server.obj.snapshot:
+            snapshot = self.get_snapshot_by_name(
+                server.obj.snapshot.rootSnapshotList, snapshot_name)
```

Comment:
I seem to remember something about some complications with this- what happens if this is called on a snapshot that already has a nested snapshot?